### ISSUE TITLE
sign cache values

### ISSUE THREAD

**Issue by Varbin** [pallets-eco/cachelib#1]
I have a suggestion: It should be possible to sign (/ apply HMAC) to cache values in the same way `werkzeug.contrib.securecookie` does already.

`pickle` is used as serializer to serialize the content. While this is absolutely fine as long nobody can access the underlying cache back end (Redis, FS, Memcached), it may allow privilege escalation once an attacker gains access to it, as `pickle` allows to store arbitrary code.

Proposal:
1. Add a warning to the documentation.
2. Add the option pass a signing key to sign the results and raise a warning if no signing key is passed at initialization.
3. Deprecate not using a signing key and ultimately enforce using one.

Practically pallets'  `ItsDangerous` could be used here.
If wanted, I can create a pull request implementing my proposal.

**Comment by northernSage** [comment]
Hey @Varbin, I would be happy to review a PR for this if you are still interested!

**Comment by Varbin** [comment]
Sure! I think #11 should be solved first, to not have accidentally conflicting code.

Anyway, my proposed change will be like:
 - Add `itsdangerous~=2.0.0` as installation requirement.
 - Add *key* kwarg to the class constructors
 - Add `_load` and `_dumps` methods to the `BaseCache` class, which will replace calls to `pickle.loads` / `pickle.dumps` - if a key is set this will call the itsdangerous instance, otherwise plain pickle.
 - Add a deprecation warning to calls without a key.

The only class that will _not_ support the key argument will be memcached, as the memcache client libraries already do serialization.

**Comment by ThiefMaster** [comment]
> Add a deprecation warning to calls without a key.

I'm not sure on that part. IMHO it should be up to the user. Not signing cache data is only an issue when you have an attacker with arbitrary write access to your cache backend - which means you are either already compromised (in that case the secret might no longer be secret anyway) or severely misconfigured stuff (redis bound to a public IP *and* not blocked by a local firewall *and* directly exposed to a malicious network/client)

**Comment by northernSage** [comment]
> Sure! I think #11 should be solved first, to not have accidentally conflicting code.

Don't let this hold you back @Varbin. If you feel like getting started, please go ahead. I'll just solve the conflicts when the time comes.

> * Add a deprecation warning to calls without

I agree with @ThiefMaster on that one. I would rather add a very conspicuous note to our (soon to be written) documentation stating that using a key is **strongly recommended**.

**Comment by davidism** [comment]
@ThiefMaster @john0isaac I'd prefer closing this as not planned.

I'm not really sure I see the use here. If an attacker has access to your cache, that's the security issue that needs to be addressed/prevented in the first place. We get junk vulnerability reports along these lines so commonly that we have a line in our [security policy](https://palletsprojects.com/security) about not reporting it:

> Insecure configuration or code in a project using our libraries. This should be reported to the relevant project instead. The machine the project uses must be assumed to be fully trusted and otherwise secure.

ItsDangerous is used to sign cookies and tokens, data that is sent to an _untrusted_ source and then returned. Your cache backend _must_ be trusted, otherwise you have bigger issues. All this does is increase time and space use, and make the code more complex.

**Comment by john0isaac** [comment]
You have a valid point but I was thinking that it won't hurt to have this as it's an optional feature.
I'm aware of the vulnerability report and the headache it caused before and still. That's why I wanted to prioritize working on the serializer.
But if you see no value in having it I won't argue with that. Let's wait to see what @ThiefMaster says.

**Comment by ThiefMaster** [comment]
I don't feel strongly about it, but at most I'd want this to be opt-in. If it can be nicely implemented as a standalone package, even better. TBH I don't think it's common to sign caches data...

**Comment by john0isaac** [comment]
> TBH I don't think it's common to sign caches data...

I also agree with this point we never signed caches at any implementation I have seen.

I have a proposal:
What if this change is a documentation change rather than a code change?

We can have a guide showing people how to use its dangerous to sign cache values instead of embedding it in code.

If you look at the implementation you can see that it's mostly editing the serializer and serializers are editable if you subclass any backend and overwrite serializer with your own.
So this is actually doable without any code change as an extension to the library.

**Comment by davidism** [comment]
I'd prefer not to document it. It's not something we want to encourage, it's not actually providing useful security. If a project needs it for some reason, I'm sure they'll know to search for a signing library and sign the values they put in the cache. I don't think there's anything special about that process that docs here would help with.
